doom-emacs/config.org
#+TITLE: My Doom Emacs Configuration
#+AUTHOR: Ziyang Lin
#+PROPERTY: header-args    :results silent  :lexical t
#+OPTIONS:  toc:2

* Installing & setup

See the [[https://github.com/hlissner/doom-emacs][doom-emacs repository]] for complete guide.

** Dependencies

#+begin_example
brew install coreutils cmake fd git llvm ripgrep pngpaste
brew tap homebrew/cask-fonts
brew cask install font-jetbrains-mono-nerd-font
#+end_example

Run =doom doctor= to find missing dependencies.

** Installing Emacs

*** macOS with Homebrew

The [[https://github.com/d12frosted/homebrew-emacs-plus][emacs-plus]] formula:
#+begin_example
brew tap d12frosted/emacs-plus
brew install emacs-plus@27 --with-modern-pen-icon --without-imagemagick
ln -s /usr/local/opt/emacs-plus@27/Emacs.app /Applications
#+end_example

The [[https://github.com/daviderestivo/homebrew-emacs-head][emacs-head]] formula:
#+begin_example
brew tap daviderestivo/emacs-head
brew install emacs-head@27
brew install --HEAD emacs-head --with-cocoa --with-xwidgets --without-imagemagick --with-modern-icon-purple --with-harfbuzz
ln -s /usr/local/opt/emacs-head@27/Emacs.app /Applications
#+end_example

Or use [[https://github.com/railwaycat/homebrew-emacsmacport][emacs-mac-port]]:
#+begin_example
brew tap railwaycat/emacsmacport
brew cask install emacs-mac
#+end_example

*Supported features:*
|--------------+------------+------------+----------------|
|              | emacs-plus | emacs-head | emacs-mac-port |
|--------------+------------+------------+----------------|
| childframe   | ok         | ok         | flickers       |
| ligatures    | no         | no         | yes            |
| compile opts | ok         | ok         | no             |
| CJK Font     | ok         | ok         | ok             |
|--------------+------------+------------+----------------|

*Update* now using =emacs-plus@28= from Homebrew. Ligature is supported.

** Installing Doom

#+begin_example sh
git clone --depth 1 https://github.com/hlissner/doom-emacs ~/.emacs.d
~/.emacs.d/bin/doom install
#+end_example
Add ~export PATH="$PATH:$HOME/.emacs.d/bin"~ to bashrc or zshrc file.
Put the Doom config file under =~/.doom.d/=.

Manually compiling some plugins:

#+begin_example
emacs -nw --eval '(vterm-module-compile) (pdf-tools-install)'
#+end_example

** Switching Emacs versions on macOS

#+begin_example sh
brew tap d12frosted/emacs-plus
brew install emacs-plus@27    # if already installed, use `brew link` instead
rm /Applications/Emacs.app    # if Emacs.app is not a symlink, add -rf flag
ln -s $(brew --prefix)/opt/emacs-plus@27/Emacs.app /Applications
doom sync
#+end_example


* Meta

** Remember to run =doom sync= after modifying in =init.el= and =packages.el=.

Use =SPC m E= to reload this file after making changes. Also =SPC q r= restarts emacs (may take a few seconds).

To show /literal/ link in org-mode, press =SPC m l t=.

** Enable lexical scoping

This slightly improves performance.
#+begin_src elisp
;; Generated by doom-emacs/config.org    -*- lexical-binding: t; -*-
#+end_src

** Detecting OS

Doom defines ~IS-WINDOWS~, ~IS-MAC~, and ~IS-LINUX~. Use a ~cond~ block to define OS specific settings.


* Doom core configuration

** Personal Information

Set name and email for GPG configuration, email, snippets, etc.
#+begin_src elisp
(setq user-full-name    "Ziyang Lin"
      user-mail-address "jakelinnzy@gmail.com")
#+end_src

** Font

Set fonts.
#+begin_src elisp
(setq doom-font
      (font-spec :family "Sarasa Mono CL" :size 13 :weight 'regular)
      ;; (font-spec :family "JetBrainsMono Nerd Font" :size 13 :weight 'regular)

      doom-variable-pitch-font
      ;; (font-spec :family "Libre Baskerville" :size 13)
      ;; (font-spec :family "Overpass" :size 13)
      (font-spec :family "ETBembo" :size 13)

      ;; Display Emoji in system font
      emojify-display-style 'unicode
      ;; NOTE Enable this to make variable pitch font size work
      mixed-pitch-set-height t)

;; This fixes CJK alignment issues on macOS
;; TODO make it also work on Linux
(cond
 (IS-MAC
  (setq doom-unicode-extra-fonts
        `("PingFang SC" ,doom-font
          "Weather Icons" "github-octicons" "FontAwesome"
          "all-the-icons" "file-icons" "Apple Color Emoji" "Menlo")
        face-font-rescale-alist
        '(("PingFang SC" . 1.25)
          ("Publico Text" . 1.1)
          ("ETBembo" . 1.15))))
 ;; TODO
 (IS-LINUX nil))
#+end_src

If the font appears too small, modify the ~face-font-rescale-alist~ variable instead of changing the ~:size~ attribute. This makes CJK fonts display at size 13, same as ~doom-variable-pitch-font~.

The items in ~doom-unicode-extra-fonts~ matches from end to beginning, so the first element (PingFang SC in this case) has the lowest priority.

*** DONE Fixing CJK Text align issue

Desired behaviour: CJK characters are displayed exactly twice as wide as ASCII characters, so things would perfectly line up.

*Update:* The current font setting in the previous block works fine. If characters are not aligned, call ~doom/reload-font~ (=SPC h r f=) fixes them.

+Haven't figured out if Emacs can achieve this. Current workaround is to run Emacs in a terminal emulator (which breaks image display), or use special fonts that achieves this (e.g. 等距更纱黑体 Sasara-Mono-SC).+ To test the display:

#+begin_example
滚滚长江东逝水 |||||浪花淘尽英雄 是非成败转头空||||||
The quick brown|||||fox jumps over the lazy dog||||||
#+end_example

See [[http://baohaojun.github.io/perfect-emacs-chinese-font.html][this blog]] for more information.

** UI

Set theme to [[file:my-doom-ayu-mirage-theme.el][my custom color theme]] based on Ayu Mirage and make comments easier to read. The theme is actually loaded afterwards.
#+begin_src elisp
;; load theme from user config directory
(add-to-list 'custom-theme-load-path doom-private-dir)
(setq doom-theme 'my-doom-ayu-mirage
      my-doom-ayu-mirage-brighter-comments t)
#+end_src

Display relative line numbers. To cycle between (none, numbers, relative numbers), use =SPC t l= (doom/toggle-line-numbers)
#+begin_src elisp
(setq display-line-numbers-type 'relative)
;; Disable line number for text mode
;; (remove-hook! '(text-mode-hook) #'display-line-numbers-mode)
#+end_src

Maximize window on startup
#+begin_src elisp
(add-to-list 'initial-frame-alist '(fullscreen . maximized))
#+end_src

Display buffer name and major mode in the window title
#+begin_src elisp
(setq frame-title-format '("" "%b [%m] - Emacs"))
#+end_src

Show an image on splash screen
#+begin_src elisp
(setq fancy-splash-image
      (concat doom-private-dir "assets/emacs-icon-200x200.png"))
#+end_src

Indent guide character
#+begin_src elisp
;; (setq indent-guide-char "│")
(setq highlight-indent-guides-method 'bitmap
      highlight-indent-guides-bitmap-function
      #'highlight-indent-guides--bitmap-line)
#+end_src

** Cursor

Possible cursor types: =box=, =hollow=, =bar=, =hbar=.
#+begin_src elisp
(let ((cursor-color-str "#ffd580"))
  (setq evil-default-cursor        (list cursor-color-str 'box)
        evil-normal-state-cursor   (list cursor-color-str 'box)
        evil-insert-state-cursor   'bar
        evil-visual-state-cursor   'hollow
        evil-operator-state-cursor #'evil-half-cursor
        evil-emacs-state-cursor    (list "white" 'box)))
;; Don't let mixed-pitch-mode change cursor
;; t or nil
(setq mixed-pitch-cursor-type    nil)
#+end_src

** Modeline

The modeline always shows "LF UTF-8", which is unnecessary.
#+begin_src elisp
(defun my/doom-modeline-conditional-buffer-encoding ()
  "Hide the encoding information when it is not LF UTF-8."
  (setq-local doom-modeline-buffer-encoding
              (unless (or (eq buffer-file-coding-system 'utf-8-unix)
                          (eq buffer-file-coding-system 'utf-8)))))

(add-hook 'after-change-major-mode-hook #'my/doom-modeline-conditional-buffer-encoding)
#+end_src

** Change some behaviours

#+begin_src elisp
(setq auto-save-default t             ;; enable autosave for file-buffers
      delete-by-moving-to-trash t     ;; sudo rm -rf /
      undo-limit (* 100 1024 1024)    ;; Raise undo limit to 100MB
      evil-want-fine-undo t           ;; More granular undo in insert state
      truncate-string-ellipsis "…"    ;; Display unicode elipsis
      ;; Don't pollute the system clipboard
      save-interprogram-paste-before-kill t
      ;; Don't compact font caches during GC, will improve performance of the modeline plugin
      inhibit-compacting-font-caches t
      ring-bell-function #'ignore     ;; Disable bell completely
      )

;; Enable soft wrapping by default (toggle with SPC t w), except in temp
;; buffers e.g. Dired, ibuffer
(add-hook! '(prog-mode-hook text-mode-hook)
           #'visual-line-mode)

;; This makes emacs-mac-port behave like any other Mac app in multiple
;; workspaces. e.g. when you are in another workspace and click Emacs' icon
;; in the Dock, you switch to the workspace Emacs is in.
;; (menu-bar-mode 1)
#+end_src

(Relatively) smooth mouse scroll.
#+begin_src elisp
(setq mouse-wheel-scroll-amount '(1 ((shift) . 5))
      mouse-wheel-progressive-speed nil
      mouse-wheel-follow-mouse t
      scroll-step 3)
#+end_src

** Don't be evil

Indentation: use 4 spaces by default.
#+begin_src elisp
(setq tab-width 4
      evil-shift-width 4)
(setq-default indent-tabs-mode nil)
#+end_src

Default settings
#+begin_src elisp
(setq evil-split-window-below t
      evil-vsplit-window-right t
      ;; :s command has the global flag by default, adding /g cancels the flag.
      evil-ex-substitute-global t
      ;; keep 5 lines from the margin
      ;; scroll-margin 5
      ;; scroll-step 1
      ;; j / k move visual lines, etc
      evil-respect-visual-line-mode nil)
#+end_src

Treat underscore as part of a word.
#+begin_src elisp
(modify-syntax-entry ?_ "w")
#+end_src

Avy: vim-easymotion for Emacs.

Use all lowercase letters plus semicolon.
#+begin_src elisp
(setq avy-keys '(?a ?s ?d ?f ?g ?h ?j ?k ?l
                    ?q ?w ?e ?r ?t ?y ?u ?i ?o ?p
                    ?z ?x ?c ?v ?b ?n ?m 59))
#+end_src
And evil-snipe is not needed.
#+begin_src elisp
(after! evil-snipe
  (evil-snipe-mode -1))
#+end_src

Enable smartparens-mode by default. Looks like it has some issues on Emacs 28. Defining another global minor mode that does not use ~turn-on-smartparens-mode~ fixes [[https://github.com/Fuco1/smartparens/issues/1052][the issue]].
#+begin_src elisp
;; In packages.el:
;;     (package! smartparens)
(require 'smartparens)
(require 'smartparens-config)
;; Enable triple backtick (markdown style quote)
(sp-local-pair '(prog-mode text-mode) "```" "```")
;; Make {<ENTER> behave sanely
(sp-with-modes '(prog-mode)
  (sp-local-pair "{" nil :post-handlers '(("||\n[i]" "RET")))
  (sp-local-pair "[" nil :post-handlers '(("||\n[i]" "RET")))
  (sp-local-pair "(" nil :post-handlers '(("||\n[i]" "RET"))))
(after! smartparens-org
  ;; Disable ' in org-mode
  (sp-local-pair 'org-mode "'" nil :actions nil))
(defun my/fix-smartparens ()
  "Enable smartparens mode"
  (define-globalized-minor-mode my/smartparens-global-mode
    smartparens-mode smartparens-mode)
  (my/smartparens-global-mode 1)
  (define-globalized-minor-mode my/show-smartparens-global-mode
    show-smartparens-mode show-smartparens-mode)
  (my/show-smartparens-global-mode 1)
  ;; (define-globalized-minor-mode my/smartparens-global-strict-mode
  ;;   smartparens-strict-mode smartparens-strict-mode)
  ;; (my/smartparens-global-strict-mode 1)
  )
(my/fix-smartparens)
#+end_src

** Completion (company-mode)

#+begin_src elisp
;; Yes I'm lazy
(setq company-minimum-prefix-length 1)
#+end_src

** Scratch buffer

#+begin_src elisp
(add-hook 'doom-scratch-buffer-hook #'text-mode)
#+end_src

* Mappings

** Use =M-h/j/k/l= to move between split windows.

Some major or minor modes may override this binding, so I decided to define a minor mode for this and make it take precedence over any other mode to get the consistent behaviour.

#+begin_src elisp
(defvar my-top-level-mode-map (make-sparse-keymap)
  "M-h/j/k/l to move between split windows.")
(map!
 (:map my-top-level-mode-map
  :nvm "M-h" #'windmove-left
  :nvm "M-j" #'windmove-down
  :nvm "M-k" #'windmove-up
  :nvm "M-l" #'windmove-right))
(define-minor-mode my-top-level-mode
  "Allows to use M-h/j/k/l to move between split windows."
  ;; The initial value
  :init-value t
  ;; Indicator for mode line
  :lighter " my-top-level"
  ;; The minor mode map
  :keymap my-top-level-mode-map)
(define-globalized-minor-mode global-my-top-level-mode
  my-top-level-mode my-top-level-mode)
;; This makes it take precedence over any other minor mode.
(add-to-list 'emulation-mode-map-alists
             `((my-top-level-mode . ,my-top-level-mode-map)))
(provide 'my-top-level-mode)
#+end_src

** Use =m= and =,= to scroll

IMO it's far more ergonomic to scroll with a single key than combinations such as =C-u and C-d=. However in some modes (e.g. Dired) this is not desired. So I manually map these for each mode that I want.
#+begin_src elisp
(map!
 :nm  "m"   #'evil-scroll-down

doom-emacs/config.el
;; Generated by doom-emacs/config.org    -*- lexical-binding: t; -*-

(setq user-full-name    "Ziyang Lin"
      user-mail-address "jakelinnzy@gmail.com")

(setq doom-font
      (font-spec :family "Sarasa Mono CL" :size 13 :weight 'regular)
      ;; (font-spec :family "JetBrainsMono Nerd Font" :size 13 :weight 'regular)

      doom-variable-pitch-font
      ;; (font-spec :family "Libre Baskerville" :size 13)
      ;; (font-spec :family "Overpass" :size 13)
      (font-spec :family "ETBembo" :size 13)

      ;; Display Emoji in system font
      emojify-display-style 'unicode
      ;; NOTE Enable this to make variable pitch font size work
      mixed-pitch-set-height t)

;; This fixes CJK alignment issues on macOS
;; TODO make it also work on Linux
(cond
 (IS-MAC
  (setq doom-unicode-extra-fonts
        `("PingFang SC" ,doom-font
          "Weather Icons" "github-octicons" "FontAwesome"
          "all-the-icons" "file-icons" "Apple Color Emoji" "Menlo")
        face-font-rescale-alist
        '(("PingFang SC" . 1.25)
          ("Publico Text" . 1.1)
          ("ETBembo" . 1.15))))
 ;; TODO
 (IS-LINUX nil))

;; load theme from user config directory
(add-to-list 'custom-theme-load-path doom-private-dir)
(setq doom-theme 'my-doom-ayu-mirage
      my-doom-ayu-mirage-brighter-comments t)

(setq display-line-numbers-type 'relative)
;; Disable line number for text mode
;; (remove-hook! '(text-mode-hook) #'display-line-numbers-mode)

(add-to-list 'initial-frame-alist '(fullscreen . maximized))

(setq frame-title-format '("" "%b [%m] - Emacs"))

(setq fancy-splash-image
      (concat doom-private-dir "assets/emacs-icon-200x200.png"))

;; (setq indent-guide-char "│")
(setq highlight-indent-guides-method 'bitmap
      highlight-indent-guides-bitmap-function
      #'highlight-indent-guides--bitmap-line)

(let ((cursor-color-str "#ffd580"))
  (setq evil-default-cursor        (list cursor-color-str 'box)
        evil-normal-state-cursor   (list cursor-color-str 'box)
        evil-insert-state-cursor   'bar
        evil-visual-state-cursor   'hollow
        evil-operator-state-cursor #'evil-half-cursor
        evil-emacs-state-cursor    (list "white" 'box)))
;; Don't let mixed-pitch-mode change cursor
;; t or nil
(setq mixed-pitch-cursor-type    nil)

(defun my/doom-modeline-conditional-buffer-encoding ()
  "Hide the encoding information when it is not LF UTF-8."
  (setq-local doom-modeline-buffer-encoding
              (unless (or (eq buffer-file-coding-system 'utf-8-unix)
                          (eq buffer-file-coding-system 'utf-8)))))

(add-hook 'after-change-major-mode-hook #'my/doom-modeline-conditional-buffer-encoding)

(setq auto-save-default t             ;; enable autosave for file-buffers
      delete-by-moving-to-trash t     ;; sudo rm -rf /
      undo-limit (* 100 1024 1024)    ;; Raise undo limit to 100MB
      evil-want-fine-undo t           ;; More granular undo in insert state
      truncate-string-ellipsis "…"    ;; Display unicode elipsis
      ;; Don't pollute the system clipboard
      save-interprogram-paste-before-kill t
      ;; Don't compact font caches during GC, will improve performance of the modeline plugin
      inhibit-compacting-font-caches t
      ring-bell-function #'ignore     ;; Disable bell completely
      )

;; Enable soft wrapping by default (toggle with SPC t w), except in temp
;; buffers e.g. Dired, ibuffer
(add-hook! '(prog-mode-hook text-mode-hook)
           #'visual-line-mode)

;; This makes emacs-mac-port behave like any other Mac app in multiple
;; workspaces. e.g. when you are in another workspace and click Emacs' icon
;; in the Dock, you switch to the workspace Emacs is in.
;; (menu-bar-mode 1)

(setq mouse-wheel-scroll-amount '(1 ((shift) . 5))
      mouse-wheel-progressive-speed nil
      mouse-wheel-follow-mouse t
      scroll-step 3)

(setq tab-width 4
      evil-shift-width 4)
(setq-default indent-tabs-mode nil)

(setq evil-split-window-below t
      evil-vsplit-window-right t
      ;; :s command has the global flag by default, adding /g cancels the flag.
      evil-ex-substitute-global t
      ;; keep 5 lines from the margin
      ;; scroll-margin 5
      ;; scroll-step 1
      ;; j / k move visual lines, etc
      evil-respect-visual-line-mode nil)

(modify-syntax-entry ?_ "w")

(setq avy-keys '(?a ?s ?d ?f ?g ?h ?j ?k ?l
                    ?q ?w ?e ?r ?t ?y ?u ?i ?o ?p
                    ?z ?x ?c ?v ?b ?n ?m 59))

(after! evil-snipe
  (evil-snipe-mode -1))

;; In packages.el:
;;     (package! smartparens)
(require 'smartparens)
(require 'smartparens-config)
;; Enable triple backtick (markdown style quote)
(sp-local-pair '(prog-mode text-mode) "```" "```")
;; Make {<ENTER> behave sanely
(sp-with-modes '(prog-mode)
  (sp-local-pair "{" nil :post-handlers '(("||\n[i]" "RET")))
  (sp-local-pair "[" nil :post-handlers '(("||\n[i]" "RET")))
  (sp-local-pair "(" nil :post-handlers '(("||\n[i]" "RET"))))
(after! smartparens-org
  ;; Disable ' in org-mode
  (sp-local-pair 'org-mode "'" nil :actions nil))
(defun my/fix-smartparens ()
  "Enable smartparens mode"
  (define-globalized-minor-mode my/smartparens-global-mode
    smartparens-mode smartparens-mode)
  (my/smartparens-global-mode 1)
  (define-globalized-minor-mode my/show-smartparens-global-mode
    show-smartparens-mode show-smartparens-mode)
  (my/show-smartparens-global-mode 1)
  ;; (define-globalized-minor-mode my/smartparens-global-strict-mode
  ;;   smartparens-strict-mode smartparens-strict-mode)
  ;; (my/smartparens-global-strict-mode 1)
  )
(my/fix-smartparens)

;; Yes I'm lazy
(setq company-minimum-prefix-length 1)

(add-hook 'doom-scratch-buffer-hook #'text-mode)

(defvar my-top-level-mode-map (make-sparse-keymap)
  "M-h/j/k/l to move between split windows.")
(map!
 (:map my-top-level-mode-map
  :nvm "M-h" #'windmove-left
  :nvm "M-j" #'windmove-down
  :nvm "M-k" #'windmove-up
  :nvm "M-l" #'windmove-right))
(define-minor-mode my-top-level-mode
  "Allows to use M-h/j/k/l to move between split windows."
  ;; The initial value
  :init-value t
  ;; Indicator for mode line
  :lighter " my-top-level"
  ;; The minor mode map
  :keymap my-top-level-mode-map)
(define-globalized-minor-mode global-my-top-level-mode
  my-top-level-mode my-top-level-mode)
;; This makes it take precedence over any other minor mode.
(add-to-list 'emulation-mode-map-alists
             `((my-top-level-mode . ,my-top-level-mode-map)))
(provide 'my-top-level-mode)

(map!
 :nm  "m"   #'evil-scroll-down
 :nm  ","   #'evil-scroll-up
 ;; m sets marker by default, move it to M
 :nm  "M"   #'evil-set-marker

 (:after info
  (:mode Info-mode
   :nm "m" #'evil-scroll-down
   :nm "," #'evil-scroll-up))

 ;; Why the f**k is this called pdf-tools not pdf
 (:after pdf-tools
  (:map pdf-view-mode-map
   :nm "m" #'pdf-view-scroll-up-or-next-page
   :nm "," #'pdf-view-scroll-down-or-previous-page)))

(map!
 ;; map j and k only in normal mode, so v10j works as expected.
 :n   "j"   #'evil-next-visual-line
 :n   "k"   #'evil-previous-visual-line
 :n   "RET" #'evil-ex-nohighlight
 ;; Home row keys jump to beginning and end of line
 :nvmo "H"   #'evil-first-non-blank
 :nvmo "L"   #'evil-end-of-line

 :nvo "f"   #'evil-avy-goto-char
 :n   "s"   #'evil-avy-goto-char-2
 ;; Use C-f/b/p/n in Insert mode
 :i   "C-p" #'previous-line
 :i   "C-n" #'next-line


 ;; company-mode for completion
 (:after company
  :i "C-x C-x" #'company-complete
  (:map company-active-map
   ;; Tab accepts completion
   "TAB" #'company-complete-selection
   [tab] #'company-complete-selection
   ;; Return always inserts newline
   "RET"    #'newline-and-indent
   [return] #'newline-and-indent))

 ;; treemacs: NERDTree-like file explorer
 ;; C-t or SPC f t to open treemacs
 :nm "C-t" #'+treemacs/toggle
 (:leader
  (:prefix ("f" . "file")
   :desc "Add project to treemacs" "T" #'treemacs-add-project-to-workspace))
 (:after treemacs
  (:map evil-treemacs-state-map
   "s-`"  #'ignore
   "p"    nil
   "p a"  #'treemacs-add-project-to-workspace
   "p d"  #'treemacs-remove-project-from-workspace
   "M"    #'treemacs-move-file
   "v"    #'treemacs-peek
   "M-h"  nil
   "M-j"  nil
   "M-k"  nil
   "M-l"  nil)))

(defun my/toggle-vterm ()
  (interactive)
  (evil-force-normal-state)
  (+vterm/toggle nil)
  ;; Scroll to bottom and enter insert state when switching into the terminal
  (when (eq major-mode 'vterm-mode)
    (evil-insert-state)))

(map!
 ;; nvim
 :nvim "s-`" #'my/toggle-vterm)

(map! :leader
      ;; By default Magit is invoked with SPC g (a/b/c...)
      ;; These bindings can be directly accessed within the Magit status buffer
      "g" nil
      :desc "Magit"                    "g"   #'magit-status
      :desc "Format buffer"            "c f" #'lsp-format-buffer
      :desc "Go to definition"         "c g" #'evil-goto-definition
      :desc "Toggle maximized window"  "t M" #'toggle-frame-maximized
      :desc "Toggle monospace font"    "t m" #'mixed-pitch-mode
      :desc "Toggle fixed wrapping"    "t v" #'visual-fill-column-mode
      :desc "Edit string at point"     "e s" #'string-edit-at-point
      :desc "Move workspace left"    "TAB <" #'+workspace/swap-left
      :desc "Move workspace right"   "TAB >" #'+workspace/swap-right
      :desc "Run last make task"       "c m" #'+make/run-last
      :desc "Run make task"            "c M" #'+make/run
      :desc "resize-mode"              "w e" #'resize-window)

(setq which-key-idle-delay 0.5
      which-key-idle-secondary-delay 0)

(setq which-key-allow-multiple-replacements t)
(after! which-key
  (pushnew!
   which-key-replacement-alist
   '(("" . "\\`+?evil[-:]?\\(?:a-\\)?\\(.*\\)") . (nil . "◂\\1"))
   '(("\\`g s" . "\\`evilem--?motion-\\(.*\\)") . (nil . "◃\\1"))))

(setq treemacs-width 25)

(after! dired
  (map!
   :map ranger-mode-map
   :nm "c"   nil
   :nm "c f" #'dired-create-empty-file
   :nm "c d" #'dired-create-directory))

(after! dired
  (require 'dired-x)
  (setq-default dired-omit-files-p t)
  (setq dired-omit-files
        "\\`[.]?#\\|\\`[.][.]?\\'\\|flycheck_.*\\.el$"))
(setq ranger-hidden-regexp
      '("^\\."
        ;; flycheck elisp
        "^flycheck_.*\\.el$"))

(setq ivy-posframe-width     130
      ivy-posframe-min-width 130
      ivy-posframe-height     18
      ivy-posframe-min-height 10)
;; (after! ivy
;;   (ivy-posframe-mode -1))

(after! ivy
  (map! :map ivy-minibuffer-map
        "C-d"     #'ivy-immediate-done
        "C-c C-c" #'ivy-immediate-done
        "<S-Return>" #'ivy-immediate-done))

(after! magit
  ;; log output of all git commands
  (setq magit-git-debug nil)
  (map! :mode magit-mode
        :g "m" #'evil-scroll-down
        :g "," #'evil-scroll-up
        :g "M" #'magit-merge
        :g "R" #'magit-remote))

(after! ibuffer
  (map! :mode ibuffer-mode
        :i "j" #'evil-next-line
        :i "k" #'evil-previous-line
        :n "?" #'describe-mode))

(setq
 ;; Where to find projects
 projectile-project-search-path '("~/repos/")
 ;; Project root patterns
 projectile-project-root-files-bottom-up '(".root" "Cargo.toml" "requirements.txt")
 ;; Don't automatically add emacs sources into project list.
 projectile-ignored-projects '("~/" "/tmp" "~/.emacs.d/.local/straight/repos/"))

(defun projectile-ignored-project-function (filepath)
  "Return t if FILEPATH is within any of `projectile-ignored-projects'"
  (or (mapcar (lambda (p) (s-starts-with-p p filepath)) projectile-ignored-projects)))

(use-package string-edit
  :commands 'string-edit-at-point)
(after! string-edit
  (add-hook 'string-edit-mode-hook
            (lambda ()
              (setq header-line-format
                    (substitute-command-keys
                     (concat "\\[string-edit-conclude] to finish, "
                             "\\[string-edit-abort] to abort."))))))

;; (use-package taskrunner
;;   :commands (ivy-taskrunner))

;; redifine the keymap to be consistent with evil
(after! resize-window
  (setq resize-window-dispatch-alist
        '((?j resize-window--enlarge-down          " Resize - Expand down" t)
          (?k resize-window--enlarge-up            " Resize - Expand up" t)
          (?l resize-window--enlarge-horizontally  " Resize - horizontally" t)
          (?h resize-window--shrink-horizontally   " Resize - shrink horizontally" t)
          (?r resize-window--reset-windows         " Resize - reset window layout" nil)
          (?w resize-window--cycle-window-positive " Resize - cycle window" nil)
          (?W resize-window--cycle-window-negative " Resize - cycle window" nil)
          (?s split-window-below " Split window horizontally" nil)
          (?v split-window-right " Split window vertically" nil)
          (?k resize-window--delete-window " Delete window" nil)
          (?K resize-window--kill-other-windows " Kill other windows (save state)" nil)
          (?y resize-window--restore-windows " (when state) Restore window configuration" nil)
          (?? resize-window--display-menu          " Resize - display menu" nil))))
;;   List of actions for `resize-window-dispatch-default.
;;   Main data structure of the dispatcher with the form:
;;   \(char function documentation match-capitals\)

;; (after! pdf-view
;;   (require 'pdf-continuous-scroll-mode)
;;   (add-hook 'pdf-view-mode-hook #'pdf-continuous-scroll-mode))

;; Don't scale up font in zen mode
(setq +zen-text-scale 0)

(setq lsp-enable-snippet t
      lsp-idle-delay 0.3
      lsp-modeline-diagnostics-message t
      lsp-modeline-diagnostics-scope :file
      ;; enable lens if server supports
      ;; lsp-lens-auto-enable t

      ;; sideline code actions & diagnostics
      lsp-ui-sideline-enable t
      lsp-ui-sideline-show-hover nil
      lsp-ui-sideline-show-code-actions t
      lsp-ui-sideline-show-diagnostics t

      ;; hover window
      lsp-ui-doc-enable 1
      lsp-ui-doc-delay 0.5

      ;; disable the large window at the bottom that appears when I insert a
      ;; function call
      lsp-signature-auto-activate t
      lsp-signature-render-documentation nil
      lsp-eldoc-enable-hover nil

      ;; improve performance by allowing to read more frequently
      ;; doom already takes care of gc
      read-process-output-max (* 5 1024 1024))

(require 'tree-sitter)
(require 'tree-sitter-langs)
(global-tree-sitter-mode)
(add-hook 'tree-sitter-after-on-hook #'tree-sitter-hl-mode)

;; (add-hook! '(text-mode-hook org-mode-hook)
;;            :append
;;            (lambda () (spell-fu-mode -1)))





(after! python
  (setq python-prettify-symbols-alist '(("lambda" . 955)))
  ;; Use Doom's format plugin instead of LSP
  (setq-hook! 'python-mode-hook
    +format-with-lsp nil
    +format-with 'black)
  (map! :mode python-mode
        :localleader
        "f" #'py-autopep8-buffer))
(after! lsp-python-ms
  (set-lsp-priority! 'mspyls 1))

(setq +latex-viewers '(skim evince sumatrapdf zathura okular pdf-tools))
(after! latex
  (defun my/apply-tex-mode-maps ()
    "Apply tex-mode maps"
    (map! :map evil-tex-mode-map
          :nm "m" #'evil-scroll-down
          :nm "," #'evil-scroll-up
          :nm "M" #'evil-scroll-down
          (:localleader
           "a" #'TeX-command-run-all
           "b" #'TeX-command-buffer)))
  (add-hook 'LaTeX-mode-hook #'my/apply-tex-mode-maps))

(after! markdown-mode
  (map! :mode markdown-mode
        :localleader
        ;; consistent with org-mode
        :desc "Toggle hiding links" "l t" #'markdown-toggle-url-hiding))
(add-hook 'markdown-mode-hook
          (lambda ()
            (hl-line-mode -1)
            (highlight-indent-guides-mode -1)))

(add-to-list 'load-path "~/.doom.d/vendor/arduino-mode")
(setq auto-mode-alist (cons '("\\.\\(pde\\|ino\\)$" . arduino-mode) auto-mode-alist))
(autoload 'arduino-mode "arduino-mode" "Arduino editing mode." t)

(setq org-directory "~/Documents/org/")

(setq org-global-properties
      ;; directory of attachments
      '(("dir" . "attach/"))
      org-ellipsis "▼" ;; ➡, ⚡, ▼, ↴, , ∞, ⬎, ⤷, ⤵
      ;; Hides *bold* /italic/ etc.
      org-hide-emphasis-markers nil)

(map! :after org
      :mode org-mode
      (:localleader
       ;; Toggle visibility of current block
       :desc "Toggle visibility of block" "v" #'org-hide-block-toggle
       ;; Execute all code blocks in the org buffer
       :desc "Execute buffer" "E" #'org-babel-execute-buffer))

(after! org
  (add-hook! '(org-mode-hook dired-mode-hook)
             #'org-download-enable))

(after! org
  ;; give titles different font size
  (custom-set-faces!
    '(org-document-title :height 1.3)
    '(org-level-1 :height 1.3)
    '(org-level-2 :height 1.2)
    '(org-level-3 :height 1.1)
    '(org-level-4 :height 1.05)
    '(org-level-5 :height 1.0)
    '(org-level-6 :height 1.0)
    '(org-level-7 :height 1.0)
    '(org-level-8 :height 1.0)
    '(org-indent  :inherit (org-hide fixed-pitch)))
  ;; make link fixed width
  (after! mixed-pitch
    (add-to-list 'mixed-pitch-fixed-pitch-faces
                 'org-link))

  ;; prettify symbols
  (defun my/prettify-org-setup ()
    (interactive)
    (setq-local prettify-symbols-alist
                '(("#+begin_src" . ?➤) ;; ➤ 🖝 ➟ ➤ ✎
                  ("#+end_src"   . ?¶) ;; ⏹
                  ("#+begin_quote" . ?»)
                  ("#+end_quote" . ?«)
                  ("#+begin_example" . ?➟)
                  ("#+end_example" . ?¶))
                prettify-symbols-unprettify-at-point 'right-edge)
    (prettify-symbols-mode 1)
    ;; hl-line-mode looks bad on source blocks that have different background
    ;; colour
    (hl-line-mode -1)
    (highlight-indent-guides-mode -1)
    (display-line-numbers-mode -1)
    ;; Limit the window width to 80 characters to make it easier to read
    (setq fill-column 100)
    (visual-fill-column-mode 1))

  (add-hook! '(org-mode-hook) #'my/prettify-org-setup #'mixed-pitch-mode))

(after! org
  (load! "vendor/org-return-dwim" doom-private-dir)
  (map! :map evil-org-mode-map  ;; this overrides evil's default binding
        :i "RET"    #'unpackaged/org-return-dwim
        :i [return] #'unpackaged/org-return-dwim))

(after! org
  (load! "vendor/org-fix-blank-lines" doom-private-dir)
  (map! :mode org-mode
        :localleader
        :desc "Fix blank lines" "z" #'unpackaged/org-fix-blank-lines))

"Done."
